Applications/Spire/Source/Ui/TranslatedTableModel.cpp

```diff
@@ -0,0 +1,132 @@
+#include "Spire/Ui/TranslatedTableModel.hpp"
+#include <algorithm>
+
+using namespace boost::signals2;
+using namespace Spire;
+
+TranslatedTableModel::ScopeExit::ScopeExit(std::function<void()> f)
+  : m_f(std::move(f)) {}
+
+TranslatedTableModel::ScopeExit::~ScopeExit() {
+  m_f();
+}
+
+TranslatedTableModel::TranslatedTableModel(std::shared_ptr<TableModel> source)
+    : m_source(source),
+      m_transaction_level(0) {
+  m_index_mapping.resize(source->get_row_size());
+  std::iota(m_index_mapping.begin(), m_index_mapping.end(), 0);
+  m_source_operation_connection = m_source->connect_operation_signal(
+    [=] (const Operation& operation) {
+      visit(operation,
+        [=] (const AddOperation& add_operation) {
+          std::transform(m_index_mapping.begin(), m_index_mapping.end(),
```

Comment:
Optimize the common case, for example when a row is added to the end of the table this can be implemented in ```O(1)```.